Fix wave restarts and GEFS FHOUT/FHMAX

WalterKolczynski-NOAA opened the pull request:
# Description
Fixes some issues that were preventing wave restarts from operating correctly.

First, the wave restart files were not being correctly linked from `$DATA` to `$DATArestart`. The files are placed in the root of `$DATA` instead of in `${DATA}/WAVE_RESTART`, so now links for the individual files are created.

Second, the incorrect filenames were being searches for and copied as part of a rerun. Filenames were geared towards multigrid waves, which use the grid names, but single grid just uses a `ww3`. Since multigrid waves are deprecated in workflow and will soon be removed (#2637), these were updated only supporting the single-grid option.

These fixes allow forecast segments (and emergency restarts) to work correctly when waves are on.

Additionally, the `FHMAX` (and `FHOUT`) for perturbed GEFS members was being overwritten by `config.efcs` due to remnant code from when it was copied from the EnKF version. This interfered with the segment setting for those GEFS members.

Resolves #3001

# Type of change
- [x] Bug fix (fixes something broken)
- [ ] New feature (adds functionality)
- [ ] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO
- Does this change require an update to any of the following submodules? NO

# How has this been tested?
- S2SW forecast-only test with segments on Hercules

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have documented my code, including function, input, and output descriptions
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] This change is covered by an existing CI test or a new one has been added
- [x] I have made corresponding changes to the system documentation if necessary

WalterKolczynski-NOAA committed:
Fix wave restarts

Fixes some issues that were preventing wave restarts from operating
correctly.

First, the wave restart files were not being correctly linked from
`$DATA` to `$DATArestart`. The files are placed in the root of
`$DATA` instead of in `${DATA}/WAVE_RESTART`, so now links for the
individual files are created.

Second, the incorrect filenames were being searches for and copied
as part of a rerun. Filenames were geared towards multigrid waves,
which use the grid names, but single grid just uses a `ww3`. Since
multigrid waves are deprecated in workflow and will soon be removed
(#2637), these were updated only supporting the single-grid option.

These fixes allow forecast segments (and emergency restarts) to
work correctly when waves are on.

Resolves #3001

WalterKolczynski-NOAA commented:
ATTN: @weihuang-jedi

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -340,7 +340,7 @@ WW3_postdet() {
     local ww3_restart_file
     for ww3_grid in ${waveGRD} ; do
-      ww3_restart_file="${restart_dir}/${restart_date:0:8}.${restart_date:8:2}0000.restart.${ww3_grid}"
+      ww3_restart_file="${restart_dir}/${restart_date:0:8}.${restart_date:8:2}0000.restart.ww3"
```
This line does not have a variables `ww3_grid`, so there is no need to loop

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -361,6 +361,15 @@ WW3_postdet() {
+  # Link restart files
+  local restart_file
+  # Use restart_date if it was determined above, otherwise use initialization date
+  for (( vdate = "${restart_date:-${PDY}${cyc}}";  vdate <= forecast_end_cycle;
+         vdate = $(date --utc -d "${vdate:0:8} ${vdate:8:2} + ${restart_interval} hours" +%Y%m%d%H) )); do
+    restart_file="${vdate:0:8}.${vdate:8:2}0000.restart.ww3"
+    ${NLN} "${DATArestart}/WW3_RESTART/${restart_file}" "${restart_file}"
+  done
```
I think we should split this into 2 parts.
part 1 should be linking the restart for initialization.  This is either the restarts at `restart_date` (if found above) or `PDYcyc`.
If it is `PDYcyc`, the `restart_dir` is `COMIN_WAVE_RESTART`.  If it is a `RERUN=YES`, it is `DATArestart/WW3_RESTART`.  This restart **must** exist.

part 2 is creating the links from `restart_date+restart_interval` to `forecast_end_cycle`
These are the restarts that will be created as the model integrates.

WalterKolczynski-NOAA replied:
There's already a separate part to copy the initial restart in, if necessary. Just need to adjust the start date for the links when that happens.

aerorahul reviewed (changes requested):
I think there is an issue in the `forecast_postdet.sh` part.  Also a suggestion to break that into 2 parts.

WalterKolczynski-NOAA committed:
Correct wave restart fixes

Removes an unnecessary loop when staging wave restarts and stops
creating output links for restart files that were already copied
in as ICs.

Also updates FHMIN_WAV when running with segments to stop creating
unneeded links from earlier in the forecast period.

WalterKolczynski-NOAA commented:
Will need to add in additional fixes for GEFS perturbed members, as they are ignoring the segments.

WalterKolczynski-NOAA committed:
Fix GEFS FHMAX with segments

The gefs `config.efcs` still contained code from the enkf version
of the file that was overwriting the FHMAX set for segments.

weihuang-jedi commented:
I have test the PR code on Hercules. I worked as expected. Fixed the segments issue.
Thanks @WalterKolczynski-NOAA !

WalterKolczynski-NOAA committed:
Merge branch 'develop' into hotfix/segs

TerrenceMcGuinness-NOAA commented:
Jenkins controller had technical difficulties and was off line.  Clearing Hercules is under currently under maintenance.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into hotfix/segs

WalterKolczynski-NOAA commented:
I need to make additional changes. The ICs and stage IC job is expecting wave restarts to use the wave grid, but when not using multigrid the files should actually be `ww3`.

WalterKolczynski-NOAA committed:
Fix stage_ic and postdet branching for single wave restart

DavidHuber-NOAA commented:
> Experiment C96C48_hybatmaerosnowDA **FAILED** on Hera in Build# 1 with error logs:
> 
> ```
> /scratch1/NCEPDEV/global/CI/3009/RUNTESTS/COMROOT/C96C48_hybatmaerosnowDA_4c3ac105/logs/2021122018/gdas_aeroanlinit.log
> /scratch1/NCEPDEV/global/CI/3009/RUNTESTS/COMROOT/C96C48_hybatmaerosnowDA_4c3ac105/logs/2021122018/gfs_aeroanlinit.log
> ```
> 
> Follow link here to view the contents of the above file(s): [(link)](https://gist.github.com/emcbot/f18065fe4825d02be54c101c3f99d178)

I am receiving the same error on WCOSS2.  There is a mismatch between the `viirs_n20` name in the GDA: `gfs.tCCz.viirs_n20_aod.tm00.nc` and the naming convention in the [Jinja yaml in the GDASApp](https://github.com/NOAA-EMC/GDASApp/blob/c5ff4e7e3bd17532211d077259ba8c7e16e5ffbb/parm/aero/obs/config/viirs_n20_aod.yaml.j2#L6).

It looks like this will be fixed by https://github.com/NOAA-EMC/global-workflow/pull/2978.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into hotfix/segs

WalterKolczynski-NOAA commented:
> > Experiment C96C48_hybatmaerosnowDA **FAILED** on Hera in Build# 1 with error logs:
> > ```
> > /scratch1/NCEPDEV/global/CI/3009/RUNTESTS/COMROOT/C96C48_hybatmaerosnowDA_4c3ac105/logs/2021122018/gdas_aeroanlinit.log
> > /scratch1/NCEPDEV/global/CI/3009/RUNTESTS/COMROOT/C96C48_hybatmaerosnowDA_4c3ac105/logs/2021122018/gfs_aeroanlinit.log
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Follow link here to view the contents of the above file(s): [(link)](https://gist.github.com/emcbot/f18065fe4825d02be54c101c3f99d178)
> 
> I am receiving the same error on WCOSS2. There is a mismatch between the `viirs_n20` name in the GDA: `gfs.tCCz.viirs_n20_aod.tm00.nc` and the naming convention in the [Jinja yaml in the GDASApp](https://github.com/NOAA-EMC/GDASApp/blob/c5ff4e7e3bd17532211d077259ba8c7e16e5ffbb/parm/aero/obs/config/viirs_n20_aod.yaml.j2#L6).
> 
> It looks like this will be fixed by #2978.

This is actually an issue with the IC names for wave restarts. I made symlinks on Hera that need to by sync'd over.

WalterKolczynski-NOAA committed:
Fix logic error in copying wave restart file

WalterKolczynski-NOAA committed:
Merge branch 'develop' into hotfix/segs

WalterKolczynski-NOAA committed:
Merge branch 'develop' into hotfix/segs

DavidHuber-NOAA commented:
I'm going to manually run just the GEFS tests on Hera.

DavidHuber-NOAA commented:
All tests passed on Hera, but after a conversation with Walter and Kate, we will wait to merge this until an additional fix for the stage_ic job goes in.

KateFriedman-NOAA committed:
Update wave staging yaml

- Add elif for prior wave restart filename for
when the new filename is not available in IC set.

Refs #3001

WalterKolczynski-NOAA committed:
Swap order of wave restart check

WalterKolczynski-NOAA committed:
Update wave staging yaml for old and new restart filename

WalterKolczynski-NOAA committed:
Merge branch 'develop' into hotfix/segs

DavidHuber-NOAA committed:
Merge branch 'develop' into hotfix/segs
